### DISCUSSION TITLE
Garbage collection warnings on app startup

### DISCUSSION THREAD

**Discussion by ediskandarov-cur** [sqlalchemy/sqlalchemy#7164] [Usage Questions] [answered]
Hi there,

I'm building an application based on FastAPI and SQLAlchemy 1.4(with asyncpg driver).

And on the app startup I see the following warnings in logs:

```console
$ poetry run uvicorn example.app_gc_warning:app
INFO:     Started server process [11310]
INFO:     Waiting for application startup.
The garbage collector is trying to clean up connection <AdaptedConnection <asyncpg.connection.Connection object at 0x10c89e040>>. This feature is unsupported on async dbapi, since no IO can be performed at this stage to reset the connection. Please close out all connections when they are no longer used, calling ``close()`` or using a context manager to manage their lifetime.
./example/app_gc_warning.py:14: SAWarning: The garbage collector is trying to clean up connection <AdaptedConnection <asyncpg.connection.Connection object at 0x10c89e040>>. This feature is unsupported on async dbapi, since no IO can be performed at this stage to reset the connection. Please close out all connections when they are no longer used, calling ``close()`` or using a context manager to manage their lifetime.
  await engine.connect()
INFO:     Application startup complete.
```

Libs in use:

```
sqlalchemy==1.4.25
asyncpg==0.24.0
fastapi==0.68.2
```

My app is defined like this:

```python
from fastapi import FastAPI
from sqlalchemy.ext.asyncio import create_async_engine
import os

db_url = os.getenv("DB_URL")
engine = create_async_engine(db_url, future=True)


app = FastAPI()


@app.on_event("startup")
async def connect_database_engine() -> None:
    await engine.connect()


@app.on_event("shutdown")
async def close_database_connection_pools() -> None:
    await engine.dispose()
```

Am I doing something wrong?

**Comment by CaselIT** [comment, marked as answer]
Hi,

As the name implies, `await engine.connect()` creates a connection, and that connection is not properly closed in your code.
While when using normal sync driver, the garbage collection can properly close it, the same is not true for async drives, since the garbage collection finalyzer is a sync function

I'm not sure what you are trying to achieve by this?
```py
@app.on_event("startup")
async def connect_database_engine() -> None:
    await engine.connect()
```

**Reply by ediskandarov-cur** [reply]
I want to initiate a connection to the database on the app startup.

What's a proper way to do that?

**Reply by CaselIT** [reply]
> I want to initiate a connection to the database on the app startup.

you do not seem to be using that connection though.

> What's a proper way to do that?

you would need to then close the connection. so either use a context manager or explicitly call close on it.

**Reply by ediskandarov-cur** [reply]
for practical reasons I do not present all the code I have.

what happens next is

```python
SessionLocal = sessionmaker(
    engine, expire_on_commit=False, class_=AsyncSession, future=True
)


async def get_db() -> AsyncGenerator[AsyncSession, None]:
    async with SessionLocal() as session:
        yield session


@router.post("/users/login")
async def login(
    db: AsyncSession = Depends(get_db),
) -> Any:
    user = await db.execute(select(User)) ...
```

**Reply by ediskandarov-cur** [reply]
Do I even need to bother about explicit engine connection and disposal on app startup/shutdown?

**Reply by CaselIT** [reply]
The disposal is probably best to keep it.

The session maker will create a connection for you automatically, by calling engine.connect itself.

The only advantage you would have by using the startup event is to fail on startup instead of on the first request if there is an issue with the engine (also the first connection will execute some dialect setup doing some db queries, so it takes a bit longer and following request)

You could probably do something like this to verify the db connectivity
```py
@app.on_event("startup")
async def connect_database_engine() -> None:
    async with engine.connect() as conn:
      # verify db connectivity
      await conn.scalar(select(1))
```

**Reply by ediskandarov-cur** [reply]
Got it. Thanks for your comments 🙏 

I will keep disposal and think of converting startup event to connectivity verification.
